### PR TITLE
[FLINK-25079][table-common] Add some initial assertj assertions for table data and types apis

### PR DESCRIPTION
**Pull request by slinkydeveloper**
## What is the purpose of the change

Added some basic utilities for testing table apis, like:

* `assertThat(DataType)`
* `assertThat(LogicalType)`
* `assertThat(RowData)`

Also add `FlinkAssertions` in `flink-test-utils-junit` to add some assertions and utilities to use together with assertj. Also deprecate the old matchers for hamcrest.

## Brief change log

* Add `TableAssertions` to encapsulate all the assertion constructors for our custom assertions
* Add `FlinkAssertions`
* Refactored some tests from junit assertions and hamcrest to assertj

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -595,6 +594,13 @@ under the License.
 				<version>${junit4.version}</version>
 			</dependency>
 
+			<dependency>
+				<groupId>org.assertj</groupId>
+				<artifactId>assertj-core</artifactId>
+				<version>${assertj.version}</version>
+				<scope>test</scope>
+			</dependency>
```
Comment (fapaul) [inline pom.xml:602]: Nit: Should we make a separate PR to add this change because it is currently necessary to use assertj in any other module?

Comment (slinkydeveloper) [reply pom.xml:602]: Our dependencies already include assertj: https://github.com/apache/flink/commit/c4b604885a98aead3bf109189cbff7d5d95ade72. This change adds it to the `dependencyManagement` as well in order to let flink-test-utils-junit pick the right version

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/utils/DataTypeUtilsTest.java [hunk 2]
```diff
@@ -68,92 +66,79 @@ public void testProjectRow() {
...
+            new Condition<>(DataTypeUtils::isInternal, "internal");
+
     @Test
     public void testIsInternalClass() {
-        assertTrue(DataTypeUtils.isInternal(DataTypes.INT()));
-        assertTrue(DataTypeUtils.isInternal(DataTypes.INT().notNull().bridgedTo(int.class)));
-        assertTrue(DataTypeUtils.isInternal(DataTypes.ROW().bridgedTo(RowData.class)));
-        assertFalse(DataTypeUtils.isInternal(DataTypes.ROW()));
+        assertThat(DataTypes.INT()).is(INTERNAL);
+        assertThat(DataTypes.INT().notNull().bridgedTo(int.class)).is(INTERNAL);
```
Comment (twalthr) [inline flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/utils/DataTypeUtilsTest.java:117]: btw it could be useful to add this to `DataTypeAssert`, also the fields names

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/FlinkAssertions.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Some reusable assertions and utilities for AssertJ. */
+public final class FlinkAssertions {
+
+    private FlinkAssertions() {}
+
+    /**
+     * Shorthand for {@code * assertThat(throwable).extracting(FlinkAssertions::chainOfCauses,
```
Comment (twalthr) [inline flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/FlinkAssertions.java:38]: Remove `*`. Maybe a block code example, instead?

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/FlinkAssertions.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+            return Stream.empty();
+        }
+        if (throwable.getCause() == null) {
+            return Stream.of(throwable);
+        }
+        return Stream.concat(Stream.of(throwable), chainOfCauses(throwable.getCause()));
+    }
+
+    @SuppressWarnings({"rawtypes", "unused"})
+    public static final InstanceOfAssertFactory<Stream, ListAssert<Throwable>> STREAM_THROWABLE =
```
Comment (twalthr) [inline flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/FlinkAssertions.java:65]: constants to the top of the class and with JavaDoc.

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/FlinkMatchers.java [hunk 5]
```diff
@@ -35,7 +35,14 @@
...
 
-/** Some reusable hamcrest matchers for Flink. */
+/**
+ * Some reusable hamcrest matchers for Flink.
+ *
+ * @deprecated You should assertj assertions, which have built-in assertions for {@link
+ *     CompletableFuture}. To check chains of {@link Throwable} causes, use {@link
+ *     FlinkAssertions#assertThatChainOfCauses(Throwable)}
+ */
+@Deprecated
```
Comment (twalthr) [inline flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/FlinkMatchers.java:45]: Changes for the core Flink testing infrastructure should not happen in a `table-common` commit. Please create a separate commit.

**Review by twalthr — CHANGES_REQUESTED** [review]
Thanks for this nice improvement of our testing infrastructure. For future PRs, please try to separate the commits to make a review easier. Commit messages that contain `and` are usually a good indicator for spliting a commit. In this particular example, you could have used at least 3-4 commits: One with the new TableAssertions, one that updates existing tests, one for global FlinkAssertions and pom.xml changes, etc.

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/FlinkMatchers.java (hunk 5 above)
Comment (slinkydeveloper) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/FlinkMatchers.java:45]: Refactored in a separate commit annotated as hotfix

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/utils/DataTypeUtilsTest.java (hunk 2 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/utils/DataTypeUtilsTest.java:117]: I think in general it's a good idea to use assertj `Condition`s more than custom assert methods when there is a non parametrized property to test on a specific object, and their failure gives no particular error message.  `Condition`s tend to be more flexible, because you can negate them, combine them, exactly as Java's `Predicate`. A couple of examples from this PR that explains my criteria:

* `isJavaSerializable` is a method because the failure gives some important indication
*  `hasSerializableString` is a method because it's parametrized (hence we need its inverse)
*  `INTERNAL` is a condition, because it's just a property check without any useful error message in case the check fails

For some properties, i think it makes sense to have both a `Condition` and a method in the assert class, for other very specific ones (like this `INTERNAL`), it makes sense to have only a `Condition`.

I pushed a commit that adds two classes where we can have `Conditions` both for `DataType` and `LogicalType`.

**Comment by slinkydeveloper** [comment]
@twalthr this is ready for another pass

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/FlinkAssertions.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            new InstanceOfAssertFactory<>(Stream.class, Assertions::<Throwable>assertThat);
+
+    /**
+     * Shorthand to assert chain of causes. Same as: <code>
+     * assertThat(throwable)
+     *     .extracting(FlinkAssertions::chainOfCauses, FlinkAssertions.STREAM_THROWABLE)
+     * </code>
+     */
+    public ListAssert<Throwable> assertThatChainOfCauses(Throwable root) {
```
Comment (twalthr) [inline flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/FlinkAssertions.java:48]: method is not static

**Review by twalthr — CHANGES_REQUESTED** [review]
Thanks for the update @slinkydeveloper. I still have the feeling that one of the most common assertions for checking a cause in a complex stack trace is missing.

Something like this would be useful:
```
assertThatThrownBy(() -> runTypeInference(actualArgumentTypes))
                        .satisfies(anyCause(Class, String));
```
Maybe we can already update `InputTypeStrategiesTestBase` for that.

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/FlinkAssertions.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    /** @see #chainOfCauses(Throwable) */
+    @SuppressWarnings({"rawtypes", "unused"})
+    public static final InstanceOfAssertFactory<Stream, ListAssert<Throwable>> STREAM_THROWABLE =
+            new InstanceOfAssertFactory<>(Stream.class, Assertions::<Throwable>assertThat);
+
+    /**
+     * Shorthand to assert the chain of causes includes a {@link Throwable} matching a specific
+     * {@link Class} and containing the provided message. Same as:
+     *
+     * <pre><code>
```
Comment (twalthr) [inline flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/FlinkAssertions.java:47]: not sure if this JavaDoc style is correct. see also: https://reflectoring.io/howto-format-code-snippets-in-javadoc/

**Review by twalthr — APPROVED** [review]
LGTM, I found one minor thing that I will fix while merging.
